src/openlcb/BroadcastTimeDefs.cxx

```diff
@@ -0,0 +1,161 @@
+/** @copyright
+ * Copyright (c) 2018, Stuart W. Baker
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are  permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file BroadcastTimeDefs.cxx
+ *
+ * Static definitions for implementations of the OpenLCB Broadcast Time
+ * Protocol.
+ *
+ * @author Stuart W. Baker
+ * @date 30 July 2022
+ */
+
+#include "BroadcastTimeDefs.hxx"
+
+#include <string>
+
+namespace openlcb
+{
+
+extern "C"
+{
+// normally requires _GNU_SOURCE
+char *strptime(const char *__restrict, const char *__restrict,
+               struct tm *__restrict);
+}
+
+//
+// BroadcastTimeDefs::time_to_string()
+//
+std::string BroadcastTimeDefs::time_to_string(int hour, int min)
+{
+    if (hour < 0 || hour > 23)
+    {
+        hour = 0;
+    }
+    if (min < 0 || min > 59)
+    {
+        min = 0;
+    }
+
+    struct tm tm;
+    tm.tm_hour = hour;
+    tm.tm_min = min;
+    char value[6];
+    if (strftime(value, 6, "%R", &tm) != 0)
+    {
+        return value;
+    }
+    else
+    {
+        return "Error";
+    }
+}
+
+//
+// BroadcastTimeDefs::rate_quarters_to_string()
+//
+std::string BroadcastTimeDefs::rate_quarters_to_string(int16_t rate)
+{
+    std::string result;
+    if (rate < 0)
+    {
+        result.push_back('-');
+        rate = -rate;
+    }
+    uint16_t whole = rate >> 2;
+    uint16_t frac = rate & 0x3;
+    if (whole >= 100)
+    {
+        result.push_back('0' + (whole / 100));
+    }
+    if (whole >= 10)
+    {
+        result.push_back('0' + ((whole % 100) / 10));
+    }
+    result.push_back('0' + (whole % 10));
+    switch (frac)
+    {
+        default:
+        case 0:
+            result.append(".00");
```

Comment:
should we skip this? If the rate is an integer, rendering it as an integer makes the most sense.